controlplane/kubeadm/internal/controllers/helpers.go

```diff
@@ -466,18 +466,22 @@ func (r *KubeadmControlPlaneReconciler) computeDesiredMachine(kcp *controlplanev
 		desiredMachine.Spec.Bootstrap.ConfigRef = existingMachine.Spec.Bootstrap.ConfigRef
 		desiredMachine.Spec.ReadinessGates = existingMachine.Spec.ReadinessGates
 	}
-	ensureMandatoryReadinessGates(desiredMachine)
+	ensureMandatoryReadinessGates(kcp.Spec.ReadinessGates, desiredMachine)
 
 	return desiredMachine, nil
 }
 
-func ensureMandatoryReadinessGates(m *clusterv1.Machine) {
+func ensureMandatoryReadinessGates(kcpReadinessGates []clusterv1.MachineReadinessGate, m *clusterv1.Machine) {
+	allReadinessGates := []clusterv1.MachineReadinessGate{}
+	allReadinessGates = append(allReadinessGates, kcpReadinessGates...)
+	allReadinessGates = append(allReadinessGates, mandatoryMachineReadinessGates...)
+
```

Comment:
Shouldn't everything below this go away and we just set it (including dedup) because we use SSA?


```suggestion
	allReadinessGates := []clusterv1.MachineReadinessGate{}
	allReadinessGates = append(allReadinessGates, kcpReadinessGates...)
	allReadinessGates = append(allReadinessGates, mandatoryMachineReadinessGates...)

  setReadinessGateConditionTypes := sets.Set[string]{}
  m.Spec.ReadinessGates = []clusterv1.MachineReadinessGate{}
  for _, readinessGate := range allReadinessGates {
    if setReadinessGateConditionTypes.Has(readinessGate.ConditionType) {
      continue
    }
    setReadinessGateConditionTypes.Insert(readinessGate.ConditionType)
    m.Spec.ReadinessGates = append(m.Spec.ReadinessGates, readinessGate)
  }
```

And drop all below?

Reply:
I have refactored this, PTAL